core/sequencer/sequencer.go

```diff
@@ -174,6 +191,9 @@ func (s *Sequencer) NewReceiver(ctx context.Context, d *domain.Domain) (mutator.
 	last := time.Unix(0, int64(mapRoot.TimestampNanos))
 
 	return s.queue.NewReceiver(ctx, last, d.DomainID, func(mutations []*mutator.QueueMessage) error {
+		if err := checkMastership(); err != nil {
```

Comment:
I took some time for me to understand the callbacks logic here. What do you think about the following (more straightforward?) approach?
1. Make `Controller` aware of `Sequencer` (store a pointer to it), and call `NewReceiver` directly.
2. `Controller` calls `NewReceiver` with a cancelable `Context`.
3. `Controller` does `IsMaster` checks periodically.
4. And cancels the `Context` when it's not master anymore.

Points (2)-(4) uses what I call a "mastership context".

Reply:
I'd love to simplify the logic. 

I like all of this, the question is what to do when the receiver from `NewReceiver` gets a callback.  When the callback starts, can it know if it is still master?

Reply:
Yes, it can check the status of `Context` instead of calling `checkMastership`, or, even better, just pass this `ctx` around as usual. One more time:
1. `Controller` creates a cancelable `Context` and makes sure to close it when it discovers mastership loss.
2. It passes this context to `Sequencer.NewReceiver`, which in turn passes it to `s.queue.NewReceiver`.
3. When the receiver is invoked, it can check `ctx.Done()` explicitly, but also it can pass this `ctx` around so that all the dependent operations can be canceled. For example, the below `CreateEpoch` call can be canceled in the middle of operation.